javaslang/src/main/java/javaslang/collection/Iterator.java

```diff
@@ -1746,12 +1738,10 @@ default String stringPrefix() {
 
     @Override
     default Option<Iterator<T>> tailOption() {
-        if (hasNext()) {
+        return Option.when(hasNext(), () -> {
             next();
-            return Option.some(this);
-        } else {
-            return Option.none();
-        }
+            return this;
```

Comment:
An iterator actually internally is mutable, like in Scala. Therefore calling multiple intermediate operations might produce unexpected results (see javadoc).